src/bun.js/api/BunObject.zig

```diff
@@ -1952,7 +1952,19 @@ pub const Crypto = struct {
                     return true;
                 },
                 .bcrypt => {
-                    pwhash.bcrypt.strVerify(previous_hash, password, .{ .allocator = allocator }) catch |err| {
+                    var password_to_use = password;
+                    var outbuf: [bun.sha.SHA512.digest]u8 = undefined;
+
+                    // bcrypt silently truncates passwords longer than 72 bytes
+                    // we use SHA512 to hash the password if it's longer than 72 bytes
+                    if (password.len > 72) {
```

Comment:
☝️ This is good to know! Can the 72 be put in a constant like BCRYPT_MAX_LEN to help in possibly other areas.

Reply:
Do we know how other folks handle this situation (other projects that expose this kind of functionality)?
Some switched to [Argon2](https://en.wikipedia.org/wiki/Argon2)
https://www.monterail.com/blog/more-secure-passwords-bcrypt

Reply:
> ☝️ This is good to know! Can the 72 be put in a constant like BCRYPT_MAX_LEN to help in possibly other areas.

I copied that comment from elsewhere in the same file. The tricky thing about that constant is it's in UTF-8 bytes, not JavaScript String.length.

> Do we know how other folks handle this situation (other projects that expose this kind of functionality)? Some switched to [Argon2](https://en.wikipedia.org/wiki/Argon2): https://www.monterail.com/blog/more-secure-passwords-bcrypt

Dropping "bcrypt" might not be the worst option, particularly since it's broken in the zig library, too (it insists on a `$2b$` prefix even though `$2a$` is also seen in the wild).

Reply:
Instead of working around the limitation I'm leaning towards the suggestion of others in this situation and favor throwing until we can switch to Argon2 or other alternative.

Reply:
> Instead of working around the limitation I'm leaning towards the suggestion of others in this situation and favor throwing until we can switch to Argon2 or other alternative.

I would go one step further: unless it's made explicit by an extra options argument how to handle passwords longer than 72 bytes, print a deprecation warning. The default strategy should be to throw when encountering a long password.

Rationale: 72 is just too low a limit. Assuming our input data has significant entropy in the first 72 bytes is dangerous. For example, I could see people using JSON data (or, uhm, XML) as a "password", and the common prefix of "password" documents may reduce the actual entropy significantly. I believe SHA512-ing our input data is nonstandard, and doing so for long inputs but not short ones seems inconsistent to me. That matters because people may well do some or all of the bcrypt work client-side.

Reply:
copying what i said in discord:

> bun isnt really the sort of program to print deprecation warnings. we dont have any of these yet. idk if we want to
maybe it's a good idea, but i think we can add the flag first and preserve backwards compatibility, and then we can think about deprecation.
>
> the reason for sha512 here is briefly mentioned in this comment by the author of libsodium.
https://github.com/oven-sh/bun/pull/3204#issuecomment-1575647630

Reply:
Deprecation warnings can be a whole other beast to tackle _(Node's machinery around it is non-trivial)_. It's good to have but if we can float on it for a bit (lower pri) that's a win in my book.